src/main/mainEvents.ts

```diff
@@ -13,75 +13,140 @@ interface MainEventNames {
+   */
+  'update-network-status'(online: boolean): void;
+
+  /**
+   * Emitted when the integration state has changed.
+   *
+   * @param state A mapping of WSL distributions to the current state, or a
+   * string if there is an error.
+   */
+  'integration-update'(state: Record<string, boolean | string>): void;
+
+  /**
+   * Emitted as a request to get the credentials for API access.
+   */
+  'api-get-credentials'(): void;
+
+  /**
+   * Emitted as a reply to 'api-get-credentials'; the credentials can be used
+   * via HTTP basic auth on localhost.
+   *
+   * @note These credentials are meant for the UI; using them may require user
+   * interaction.
+   */
+  'api-credentials'(credentials: { user: string, password: string, port: number }): void;
 }
 
+type HandlerParams<eventName extends keyof MainEventNames> =
+  void extends ReturnType<MainEventNames[eventName]> ? never :
+  Parameters<MainEventNames[eventName]>;
+
+type HandlerReturn<eventName extends keyof MainEventNames> =
+  void extends ReturnType<MainEventNames[eventName]> ? never :
+  ReturnType<MainEventNames[eventName]>;
+
+type HandlerType<eventName extends keyof MainEventNames> =
+  void extends ReturnType<MainEventNames[eventName]> ? never :
+  (...args: HandlerParams<eventName>) => Promise<HandlerReturn<eventName>>;
+
 interface MainEvents extends EventEmitter {
   emit<eventName extends keyof MainEventNames>(
-    event: eventName,
+    event: void extends ReturnType<MainEventNames[eventName]> ? eventName : never,
     ...args: Parameters<MainEventNames[eventName]>
   ): boolean;
-  /* @deprecated */
+  /** @deprecated */ // Deprecate the untyped form, to prevent typos.
   emit(eventName: string | symbol, ...args: any[]): boolean;
   on<eventName extends keyof MainEventNames>(
-    event: eventName,
+    event: void extends ReturnType<MainEventNames[eventName]> ? eventName : never,
     listener: (...args: Parameters<MainEventNames[eventName]>) => void
   ): this;
-  /* @deprecated */
+  /** @deprecated */ // Deprecate the untyped form, to prevent typos.
   on(event: string | symbol, listener: (...args: any[]) => void): this;
+
+  /**
+   * Invoke a handler that will (eventually) return a result.
+   */
+  invoke<eventName extends keyof MainEventNames>(
+    event: void extends ReturnType<MainEventNames[eventName]> ? never : eventName,
+    ...args: HandlerParams<eventName>): Promise<HandlerReturn<eventName>>;
+
+  /**
+   * Register a handler that will handle invoke() callers.
+   */
+  handle<eventName extends keyof MainEventNames>(
+    event: void extends ReturnType<MainEventNames[eventName]> ? never : eventName,
+    handler: HandlerType<eventName>
+  ): void;
+}
+
+class MainEventsImpl extends EventEmitter implements MainEvents {
+  handlers: {
+    [eventName in keyof MainEventNames]?: HandlerType<eventName> | undefined;
+  } = {};
+
+  async invoke<eventName extends keyof MainEventNames>(
```

Comment:
Another possibly dumb question -- what do these `invoke` and `handle` methods replace? Because I don't see anything that's been removed, but this code affects existing event names, not just the new diagnostic-related ones.